source/iofs/iofs.go

```diff
@@ -0,0 +1,164 @@
+// +build go1.16
+
+package iofs
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"path"
+	"strconv"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+func init() {
+	source.Register("iofs", &IoFS{})
+}
+
+// IoFS is a source driver for io/fs#FS.
+type IoFS struct {
+	migrations *source.Migrations
+	fsys       fs.FS
+	path       string
+}
+
+// Open by url does not supported with Iofs.
+func (i *IoFS) Open(url string) (source.Driver, error) {
+	return nil, errors.New("iofs driver does not support open by url")
+}
+
+// WithInstance wraps io/fs#FS as source.Driver.
+func WithInstance(fsys fs.FS, path string) (source.Driver, error) {
+	var i IoFS
+	if err := i.Init(fsys, path); err != nil {
+		return nil, fmt.Errorf("failed to init driver with path %s: %w", path, err)
+	}
+	return &i, nil
+}
+
+// Init prepares not initialized Iofs instance to read migrations from a
+// fs.ReadDirFS instance and a relative path.
+func (p *IoFS) Init(fsys fs.FS, path string) error {
+	entries, err := fs.ReadDir(fsys, path)
+	if err != nil {
+		return err
+	}
+
+	ms := source.NewMigrations()
+	for _, e := range entries {
+		if e.IsDir() {
+			continue
+		}
+		m, err := source.DefaultParse(e.Name())
+		if err != nil {
+			continue
+		}
+		file, err := e.Info()
+		if err != nil {
+			continue
```

Comment:
Is continue OK like any other error case?

Reply:
According to https://tip.golang.org/pkg/io/fs/#DirEntry an error in this case may mean:

> If the file has been removed or renamed since the directory read, Info may return an error satisfying errors.Is(err, ErrNotExist).

But I think it can also just represent any other IO error, in e.g. the case the FS interface is backed by a real FS and not `embed`. I think ignoring the error and `continue`ing is wrong here, instead any IO error should be reported back to the caller of the function I think (also in case of `DefaultParse` error above)

Reply:
Agreed that errors should be returned and not ignored. Ignored errors will make bugs harder to debug.